### ISSUE TITLE
ActivityPub support

### ISSUE THREAD

**Issue by strugee** [pump-io/pump.io#1241]
ActivityPub support is now being developed in the `activitypub` branch.

Copypasta from a list of general ActivityPub implementation tasks that I wrote:

Have we implemented URLs for:

- [ ] Object `id` properties? (Section 3.2 Retrieving objects)
- [ ] Actor inboxes?
- [ ] Actor outboxes?
- [ ] Actor Following collections?
- [ ] Actor Followers collections?
- [ ] Actor Likes collections? (optional)
- [ ] Binary uploads? (Section 6 Uploading Media, optional)

Do we handle:

- [ ] The special Public collection?
- [ ] Local Activity side effects (including Undo side effects)?
- [ ] Server-to-server Activity side effects?
- [ ] Linked Data Notifications RDF representations (very optional)? (Section 8.1 Delivery)

Do we:

- [ ] Properly support Actor simplified login naming? (Section 4)
- [ ] Deliver activities properly?
- [ ] Implement authentication and authorization checks for clients?
- [ ] Do the same for server-to-server interactions?
- [ ] Scrub Activities for cross-site scripting attacks?
- [ ] Have a plan for combating spam?

In addition to the generic AP stuff, we have an existing network/project to worry about. So, we also need to:

- [ ] Handle AS2 vs. AS1 differences
- [ ] Handle S2S differences? I've forgotten what the plan on this is to be honest
- [ ] Migrate the web UI to understanding AS2
- [ ] Migrate the web UI to using AP endpoints
- [ ] Test everything
- [ ] Run the AP test suite against our implementation

**Comment by jankusanagi** [comment]
Looking forward to this =)

**Comment by oshepherd** [comment]
WRT ID properties: I'd suggest leaving them as is and content negotiating on the existing endpoints. AP clients are required to set their Accept header when doing a GET request such that "application/ld+json; profile="https://www.w3.org/ns/activitystreams#"" is negotiated; and you can of course identify ActivityPub posts by their content type.

This permits clients to do "transparent" upgrades: They can start requesting AS2 content, and will migrate to AP as and when the servers do, without any need to rewrite URIs, etc.

**Comment by strugee** [comment]
@oshepherd yeah, I agree. Unless any unexpected issues come up that's probably going to be the way this goes.

**Comment by snarfed** [comment]
definitely looking forward to this! I've been sketching designs to bridge indieweb to ostatus and activitypub, a la https://brid.gy (some background in https://indieweb.org/2017/ostatusbridge ), and I hope to include pump.io.

**Comment by clacke** [comment]
Duplicated by: https://framagit.org/compa/compa/issues/4

**Comment by evanp** [comment]
I think we're doing all this in the as2 branch now.

**Comment by evanp** [comment]
Could we convert this to a project on Github? Every one of these checkboxes is a big effort and probably deserves its own issue. They could at least be assigned to different people.

**Comment by PopeRigby** [comment]
Is this still in progress?

**Comment by resynth1943** [comment]
I feel like you've run into this a bit quickly, and you've ended up getting tired with it.
Any way I can help?

**Comment by vxcamiloxv** [comment]
@resynth1943  we waiting the 6.0.0 release to start implement the remaining parts cc: @strugee
